File: core/src/main/java/com/tuuzed/androidx/recyclerview/adapter/CommonItemViewHolder.java
package com.tuuzed.androidx.recyclerview.adapter;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.util.SparseArray;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import androidx.annotation.DrawableRes;
import androidx.annotation.IdRes;
import androidx.annotation.IntDef;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.recyclerview.widget.RecyclerView;

public class CommonItemViewHolder extends RecyclerView.ViewHolder {

    @NonNull
    private final SparseArray<View> mViewHolder;

    public CommonItemViewHolder(View itemView) {
        super(itemView);
        mViewHolder = new SparseArray<>();
    }

    public CommonItemViewHolder(View itemView, int initialCapacity) {
        super(itemView);
        mViewHolder = new SparseArray<>(initialCapacity);
    }

    @IntDef({View.VISIBLE, View.INVISIBLE, View.GONE})
    @Retention(RetentionPolicy.SOURCE)
    @interface Visibility {
    }

    @NonNull
    public CommonItemViewHolder visibility(@IdRes int id, @Visibility int visibility) {
        find(id, View.class).setVisibility(visibility);
        return this;
    }

    @NonNull
    public CommonItemViewHolder text(@IdRes int id, @NonNull CharSequence text) {
        find(id, TextView.class).setText(text);
        return this;
    }

    @NonNull
    public CommonItemViewHolder text(@IdRes int id, @NonNull CharSequence text, TextView.BufferType type) {
        find(id, TextView.class).setText(text, type);
        return this;
    }

    @NonNull
    public CommonItemViewHolder text(@IdRes int id, char[] text, int start, int len) {
        find(id, TextView.class).setText(text, start, len);
        return this;
    }

    @NonNull
    public CommonItemViewHolder text(@IdRes int id, @StringRes int resId) {
        find(id, TextView.class).setText(resId);
        return this;
    }

    @NonNull
    public CommonItemViewHolder text(@IdRes int id, @StringRes int resId, TextView.BufferType type) {
        find(id, TextView.class).setText(resId, type);
        return this;
    }

    @NonNull
    public CommonItemViewHolder image(@IdRes int id, @DrawableRes int resId) {
        find(id, ImageView.class).setImageResource(resId);
        return this;
    }

    @NonNull
    public CommonItemViewHolder image(@IdRes int id, @NonNull Drawable drawable) {
        find(id, ImageView.class).setImageDrawable(drawable);
        return this;
    }

    @NonNull
    public CommonItemViewHolder image(@IdRes int id, @NonNull Bitmap bitmap) {
        find(id, ImageView.class).setImageBitmap(bitmap);
        return this;
    }

    @NonNull
    public CommonItemViewHolder click(@IdRes int id, View.OnClickListener listener) {
        find(id).setOnClickListener(listener);
        return this;
    }


    @NonNull
    public CommonItemViewHolder longClick(@IdRes int id, View.OnLongClickListener listener) {
        find(id).setOnLongClickListener(listener);
        return this;
    }

    public <T extends View> T find(@IdRes int id) {
        View view = mViewHolder.get(id);
        if (view == null) {
            view = itemView.findViewById(id);
            if (view == null) throw new IllegalArgumentException("find# id not find");
            mViewHolder.put(id, view);
        }
        //noinspection unchecked
        return (T) view;
    }

    public <T extends View> T find(@IdRes int id, @NonNull Class<T> type) {
        return find(id);
    }
}

File: core/src/main/java/com/tuuzed/androidx/recyclerview/adapter/CommonItemViewHolder.kt
package com.tuuzed.androidx.recyclerview.adapter

import android.content.res.ColorStateList
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.os.Build
import android.util.SparseArray
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.*
import androidx.recyclerview.widget.RecyclerView

class CommonItemViewHolder : RecyclerView.ViewHolder {

    private val mViewHolder: SparseArray<View>

    constructor(itemView: View) : super(itemView) {
        mViewHolder = SparseArray()
    }

    constructor(itemView: View, initialCapacity: Int) : super(itemView) {
        mViewHolder = SparseArray(initialCapacity)
    }

    @IntDef(View.VISIBLE, View.INVISIBLE, View.GONE)
    @Retention(AnnotationRetention.SOURCE)
    internal annotation class Visibility

    fun visibility(@IdRes id: Int, @Visibility visibility: Int): CommonItemViewHolder {
        find(id, View::class.java).visibility = visibility
        return this
    }


    fun text(@IdRes id: Int, text: CharSequence): CommonItemViewHolder {
        findTextView(id).text = text
        return this
    }

    fun text(@IdRes id: Int, text: CharSequence, type: TextView.BufferType): CommonItemViewHolder {
        findTextView(id).setText(text, type)
        return this
    }

    fun text(@IdRes id: Int, text: CharArray, start: Int, len: Int): CommonItemViewHolder {
        findTextView(id).setText(text, start, len)
        return this
    }

    fun text(@IdRes id: Int, @StringRes resId: Int): CommonItemViewHolder {
        findTextView(id).setText(resId)
        return this
    }

    fun text(@IdRes id: Int, @StringRes resId: Int, type: TextView.BufferType): CommonItemViewHolder {
        findTextView(id).setText(resId, type)
        return this
    }

    fun textColor(@IdRes id: Int, @ColorInt color: Int): CommonItemViewHolder {
        findTextView(id).setTextColor(color)
        return this
    }

    fun textColor(@IdRes id: Int, colors: ColorStateList): CommonItemViewHolder {
        findTextView(id).setTextColor(colors)
        return this
    }

    fun background(@IdRes id: Int, background: Drawable): CommonItemViewHolder {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            find(id, View::class.java).background = background
        } else {
            @Suppress("DEPRECATION")
            find(id, View::class.java).setBackgroundDrawable(background)
        }
        return this
    }

    fun backgroundResource(@IdRes id: Int, @DrawableRes resId: Int): CommonItemViewHolder {
        find(id, View::class.java).setBackgroundResource(resId)
        return this
    }

    fun backgroundColor(@IdRes id: Int, @ColorInt color: Int): CommonItemViewHolder {
        find(id, View::class.java).setBackgroundColor(color)
        return this
    }

    fun image(@IdRes id: Int, @DrawableRes resId: Int): CommonItemViewHolder {
        findImageView(id).setImageResource(resId)
        return this
    }

    fun image(@IdRes id: Int, drawable: Drawable): CommonItemViewHolder {
        findImageView(id).setImageDrawable(drawable)
        return this
    }

    fun image(@IdRes id: Int, bitmap: Bitmap): CommonItemViewHolder {
        findImageView(id).setImageBitmap(bitmap)
        return this
    }

    fun click(@IdRes id: Int, onClick: OnClick?): CommonItemViewHolder {
        if (onClick == null) {
            find<View>(id).setOnClickListener(null)
        } else {
            find<View>(id).setOnClickListener { onClick(it) }
        }
        return this
    }

    fun longClick(@IdRes id: Int, onLongClick: OnLongClick?): CommonItemViewHolder {
        if (onLongClick == null) {
            find<View>(id).setOnLongClickListener(null)
        } else {
            find<View>(id).setOnLongClickListener { onLongClick(it) }
        }
        return this
    }

    fun <T : View> find(@IdRes id: Int, type: Class<T>): T = find(id)

    fun <T : View> find(@IdRes id: Int): T {
        var view: View? = mViewHolder.get(id)
        if (view == null) {
            view = itemView.findViewById(id)
            if (view == null) throw IllegalArgumentException("find# id not find")
            mViewHolder.put(id, view)
        }
        @Suppress("UNCHECKED_CAST")
        return view as T
    }


    private fun findTextView(@IdRes id: Int) = find(id, TextView::class.java)
    private fun findImageView(@IdRes id: Int) = find(id, ImageView::class.java)
}